compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs

```diff
@@ -76,10 +76,11 @@ pub fn finalize<'ll, 'tcx>(cx: &CodegenCx<'ll, 'tcx>) {
         let coverage_mapping_buffer = llvm::build_byte_buffer(|coverage_mapping_buffer| {
             mapgen.write_coverage_mapping(expressions, counter_regions, coverage_mapping_buffer);
         });
-        debug_assert!(
-            !coverage_mapping_buffer.is_empty(),
-            "Every `FunctionCoverage` should have at least one counter"
-        );
+
```

Comment:
When we talked about this yesterday, I suggested we keep the assert (but as `assert!`, not `debug_assert!`), and change `add_unused_functions()` to check for no coverage before adding an unused function stub:

```rust
if !tcx.covered_code_regions(def_id).is_empty() { 
        cx.define_unused_fn(non_codegenned_def_id);
} 
```

Your change will still define unused functions that you are ultimately not going to include in the coverage map. That is wasteful at the least, and possibly exposes us to other issues we're not yet aware of.

Your change also tries to permit functions that have no coverage counters but are still "used".

So far I'm not aware that we've ever run into that situation. I feel like we should not permit it, but the `assert!` will catch it if it ever happens in the future, and we can deal with it properly if it does.

Reply:
Ah, yeah you're right. I was thinking just in terms of the coverage mapping data but we really shouldn't be defining the functions at all if we're not going to use them in code coverage. I've pushed a commit that fixes that.